spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/integration/inbound/PubSubMessageSource.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2017-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.pubsub.integration.inbound;
+
+import java.util.ArrayDeque;
+import java.util.List;
+import java.util.Map;
+
+import org.springframework.cloud.gcp.pubsub.core.subscriber.PubSubSubscriberOperations;
+import org.springframework.cloud.gcp.pubsub.integration.AckMode;
+import org.springframework.cloud.gcp.pubsub.integration.PubSubHeaderMapper;
+import org.springframework.cloud.gcp.pubsub.support.GcpPubSubHeaders;
+import org.springframework.cloud.gcp.pubsub.support.converter.ConvertedAcknowledgeablePubsubMessage;
+import org.springframework.integration.IntegrationMessageHeaderAccessor;
+import org.springframework.integration.endpoint.AbstractFetchLimitingMessageSource;
+import org.springframework.integration.endpoint.AbstractMessageSource;
+import org.springframework.integration.mapping.HeaderMapper;
+import org.springframework.messaging.Message;
+import org.springframework.util.Assert;
+
+/**
+ * A <a href="https://cloud.google.com/pubsub/docs/pull#pubsub-pull-messages-sync-java">PubSub Synchronous pull</a>
+ * implementation of {@link AbstractMessageSource}.
+ *
+ * <p>Enables more even load balancing with polling adapters, at the cost of efficiency gained from streaming pull.
```

Comment:
I don't know if it's entirely accurate to say that there is always an efficiency cost with sync pull.
If you only need one message, sync pulling should be more efficient because you're not transferring a whole batch of messages.
It might be more accurate to say that message receive throughput is lower with sync.
```suggestion
 * <p>Enables more even load balancing with polling adapters, at the cost of lower maximum message receive throughput than achieved by streaming pull.
```

Reply:
I am removing this comment entirely. The docs go into enough detail about why you may use one over the other.